packages/grafana-data/src/transformations/transformers/joinDataFrames.ts

```diff
@@ -0,0 +1,294 @@
+import { DataFrame, Field, FieldMatcher, FieldType, Vector } from '../../types';
+import { ArrayVector } from '../../vector';
+import { fieldMatchers } from '../matchers';
+import { FieldMatcherID } from '../matchers/ids';
+import { getTimeField, sortDataFrame } from '../../dataframe';
+import { getFieldDisplayName } from '../../field';
+
+export function pickBestJoinField(data: DataFrame[]): FieldMatcher {
+  const { timeField } = getTimeField(data[0]);
+  if (timeField) {
+    return fieldMatchers.get(FieldMatcherID.firstTimeField).get({});
+  }
+  let common: string[] = [];
+  for (const f of data[0].fields) {
+    if (f.type === FieldType.number) {
+      common.push(f.name);
+    }
+  }
+
+  for (let i = 1; i < data.length; i++) {
+    const names: string[] = [];
+    for (const f of data[0].fields) {
+      if (f.type === FieldType.number) {
+        names.push(f.name);
+      }
+    }
+    common = common.filter((v) => !names.includes(v));
+  }
+
+  return fieldMatchers.get(FieldMatcherID.byName).get(common[0]);
+}
+
+/**
+ * @alpha
+ */
+export interface JoinOptions {
+  /**
+   * The input fields
+   */
+  frames: DataFrame[];
+
+  /**
+   * The field to join -- frames that do not have this field will be droppped
+   */
+  joinBy?: FieldMatcher;
+
+  /**
+   * Optionally filter the non-join fields
+   */
+  keep?: FieldMatcher;
+
+  /**
+   * When the result is a single frame, this will to a quick check to see if the values are sorted,
+   * and sort if necessary.  If the first/last values are in order the whole vector is assumed to be
+   * sorted
+   */
+  enforceSort?: boolean;
+
+  /**
+   * @internal -- used when we need to keep a reference to the original frame/field index
+   */
+  keepOriginIndices?: boolean;
+}
+
+/**
+ * This will return a single frame joined by the first matching field.  When a join field is not specified,
+ * the default will use the first time field
+ */
+export function outerJoinDataFrames(options: JoinOptions): DataFrame | undefined {
+  if (!options.frames?.length) {
+    return undefined;
+  }
+  const joinFieldMatcher = options.joinBy ?? pickBestJoinField(options.frames);
```

Comment:
can we avoid making this here and just declare it with `let`. if there is only 1 frame and `enforceSort` is absent (the typical case), then this doesnt get used at all.